compiler/src/compiler/mir_optimize/data_flow.rs

```diff
@@ -0,0 +1,430 @@
+                FlowValue::Symbol("Int".to_string()),
+                FlowValue::Symbol("Text".to_string()),
+                FlowValue::Symbol("Symbol".to_string()),
+                FlowValue::Symbol("List".to_string()),
+                FlowValue::Symbol("Struct".to_string()),
+                FlowValue::Symbol("Function".to_string()),
+                FlowValue::Symbol("Builtin".to_string()),
+                FlowValue::Symbol("SendPort".to_string()),
+                FlowValue::Symbol("ReceivePort".to_string()),
+            ],
+        }
+    }
+}
+
+struct Multiverse {
+    timelines: Vec<Timeline>,
+}
+impl Multiverse {
+    fn big_bang() -> Self {
+        Self {
+            timelines: vec![Timeline::default()],
+        }
+    }
+
+    fn run(&mut self, id: Id, expression: &Expression) {
+        let mut new_timelines = vec![];
+        for timeline in &self.timelines {
+            new_timelines.extend(timeline.run(id, expression));
+        }
+        self.timelines = new_timelines;
+    }
+
+    fn possible_values_for(&self, id: Id) -> FxHashSet<FlowValue> {
+        self.timelines
+            .iter()
+            .fold(FxHashSet::default(), |mut set, timeline| {
+                set.insert(timeline.values[&id].clone());
+                set
+            })
+    }
+
+    fn purge(&mut self, ids_to_purge: &[Id]) {
+        for timeline in &mut self.timelines {
+            for purged in ids_to_purge {
+                timeline.values.remove(purged);
+            }
+        }
+
+        let new_timelines = self
+            .timelines
+            .drain(..)
+            .collect::<FxHashSet<_>>()
+            .into_iter()
+            .collect();
+        self.timelines = new_timelines;
+    }
+}
+
+impl Mir {
+    pub fn gather_data_flow_insights(&self) {
+        println!("Gathering data flow insights");
+
+        let mut id_to_its_last_usage = FxHashMap::default();
+        self.body.visit_in_analysis_order(&mut |id, expression| {
+            id_to_its_last_usage.insert(id, id);
+            for referenced in expression.ids_used_for_analysis() {
+                id_to_its_last_usage.insert(referenced, id);
+            }
+        });
+
+        let mut id_to_last_usages_it_does: FxHashMap<Id, Vec<Id>> = FxHashMap::default();
+        for (used_id, usage) in id_to_its_last_usage {
+            id_to_last_usages_it_does
+                .entry(usage)
+                .or_default()
+                .push(used_id);
+        }
+
+        let mut insights: FxHashMap<Id, FxHashSet<FlowValue>> = FxHashMap::default();
+        let mut multiverse = Multiverse::big_bang();
+        self.body.visit_in_analysis_order(&mut |id, expression| {
+            multiverse.run(id, expression);
+            insights.insert(id, multiverse.possible_values_for(id));
+            multiverse.purge(id_to_last_usages_it_does.get(&id).unwrap_or(&vec![]));
+
+            info!(
+                "{id:?}: {:30}  ({} timelines)",
+                insights[&id]
```

Comment:
You can remove the map lookup for the value we just inserted:

```suggestion
            let possible_values_entry = insights
                .entry(id)
                .insert_entry(multiverse.possible_values_for(id));
            let possible_values = possible_values_entry.get();
            multiverse.purge(id_to_last_usages_it_does.get(&id).unwrap_or(&vec![]));

            info!(
                "{id:?}: {:30}  ({} timelines)",
                possible_values
```